`.getProgramDiagnostics()` is missing in the API

With the following `tsconfig.json`:

```json
{
  "compilerOptions": {
    "moduleResolution": "bundler",
    "module": "nodenext"
  }
}
```

The `tsgo` command reports:

```shell
tsconfig.json:3:23 - error TS5095: Option 'bundler' can only be used when 'module' is set to 'preserve', 'commonjs', or 'es2015' or later.
tsconfig.json:3:23 - error TS5109: Option 'moduleResolution' must be set to 'NodeNext' (or left unspecified) when option 'module' is set to 'NodeNext'.
```

When using Strada programatically, these diagnostics are returned from `program.getOptionsDiagnostics()` (not as config parse diagnostics).

Seems like Corsa is missing a similar API. I tried `program.getConfigFileParsingDiagnostics()`, `program.getSyntacticDiagnostics()`, `program.getSemanticDiagnostics()`, `program.getDeclarationDiagnostics()`, but none of them return these diagnostics.

A search of the code base for `OptionsDiagnostics` yields no results. Perhaps it would be worth adding `program.getProgramDiagnostics()` or even `project.getProjectDiagnostics()`:

```ts
const project = snapshot.getProject("./tsconfig.json");
const projectDiagnostics = project.getProjectDiagnostics();

if (projectDiagnostics.length > 0) {
	// report errors
}
```

--- a/_packages/native-preview/src/api/async/api.ts
+++ b/_packages/native-preview/src/api/async/api.ts
@@ -517,6 +517,19 @@ export class Program {
     }
 
     /**
+     * Get binder diagnostics for a specific file or all files.
+     * @param file - Optional file to get diagnostics for. If omitted, returns diagnostics for all files.
+     */
+    async getBindDiagnostics(file?: DocumentIdentifier): Promise<readonly Diagnostic[]> {
+        const data = await this.client.apiRequest<Diagnostic[]>("getBindDiagnostics", {
+            snapshot: this.snapshotId,
+            project: this.projectId,
+            ...(file !== undefined ? { file } : {}),
+        });
+        return data ?? [];
+    }
+
+    /**
      * Get semantic (type-check) diagnostics for a specific file or all files.
      * @param file - Optional file to get diagnostics for. If omitted, returns diagnostics for all files.
      */
@@ -551,6 +564,28 @@ export class Program {
             snapshot: this.snapshotId,
             project: this.projectId,
             ...(file !== undefined ? { file } : {}),
+        });
+        return data ?? [];
+    }
+
+    /**
+     * Get program-wide diagnostics for the project, including compiler options diagnostics.
+     */
+    async getProgramDiagnostics(): Promise<readonly Diagnostic[]> {
+        const data = await this.client.apiRequest<Diagnostic[]>("getProgramDiagnostics", {
+            snapshot: this.snapshotId,
+            project: this.projectId,
+        });
+        return data ?? [];
+    }
+
+    /**
+     * Get global (non-file-specific) semantic diagnostics for the project.
+     */
+    async getGlobalDiagnostics(): Promise<readonly Diagnostic[]> {
+        const data = await this.client.apiRequest<Diagnostic[]>("getGlobalDiagnostics", {
+            snapshot: this.snapshotId,
+            project: this.projectId,
         });
         return data ?? [];
     }

--- a/_packages/native-preview/src/api/sync/api.ts
+++ b/_packages/native-preview/src/api/sync/api.ts
@@ -525,6 +525,19 @@ export class Program {
     }
 
     /**
+     * Get binder diagnostics for a specific file or all files.
+     * @param file - Optional file to get diagnostics for. If omitted, returns diagnostics for all files.
+     */
+    getBindDiagnostics(file?: DocumentIdentifier): readonly Diagnostic[] {
+        const data = this.client.apiRequest<Diagnostic[]>("getBindDiagnostics", {
+            snapshot: this.snapshotId,
+            project: this.projectId,
+            ...(file !== undefined ? { file } : {}),
+        });
+        return data ?? [];
+    }
+
+    /**
      * Get semantic (type-check) diagnostics for a specific file or all files.
      * @param file - Optional file to get diagnostics for. If omitted, returns diagnostics for all files.
      */
@@ -559,6 +572,28 @@ export class Program {
             snapshot: this.snapshotId,
             project: this.projectId,
             ...(file !== undefined ? { file } : {}),
+        });
+        return data ?? [];
+    }
+
+    /**
+     * Get program-wide diagnostics for the project, including compiler options diagnostics.
+     */
+    getProgramDiagnostics(): readonly Diagnostic[] {
+        const data = this.client.apiRequest<Diagnostic[]>("getProgramDiagnostics", {
+            snapshot: this.snapshotId,
+            project: this.projectId,
+        });
+        return data ?? [];
+    }
+
+    /**
+     * Get global (non-file-specific) semantic diagnostics for the project.
+     */
+    getGlobalDiagnostics(): readonly Diagnostic[] {
+        const data = this.client.apiRequest<Diagnostic[]>("getGlobalDiagnostics", {
+            snapshot: this.snapshotId,
+            project: this.projectId,
         });
         return data ?? [];
     }

--- a/_packages/native-preview/test/async/api.test.ts
+++ b/_packages/native-preview/test/async/api.test.ts
@@ -3438,6 +3438,86 @@ describe("Program - diagnostics", () => {
             await api.close();
         }
     });
+
+    test("getBindDiagnostics", async () => {
+        const source = `let x = 1;\nlet x = 2;`;
+        const api = spawnAPI({
+            "/tsconfig.json": "{}",
+            "/src/index.ts": source,
+        });
+        try {
+            const snapshot = await api.updateSnapshot({ openProject: "/tsconfig.json" });
+            const project = snapshot.getProject("/tsconfig.json")!;
+            const diags = await project.program.getBindDiagnostics("/src/index.ts");
+            assert.deepEqual(diags, [
+                {
+                    fileName: "/src/index.ts",
+                    ...rangeOf(source, "x", 0),
+                    code: 2451,
+                    category: DiagnosticCategory.Error,
+                    text: "Cannot redeclare block-scoped variable 'x'.",
+                },
+                {
+                    fileName: "/src/index.ts",
+                    ...rangeOf(source, "x", 1),
+                    code: 2451,
+                    category: DiagnosticCategory.Error,
+                    text: "Cannot redeclare block-scoped variable 'x'.",
+                },
+            ]);
+        }
+        finally {
+            await api.close();
+        }
+    });
+
+    test("getProgramDiagnostics", async () => {
+        const config = `{ "compilerOptions": { "moduleResolution": "bundler", "module": "nodenext" } }`;
+        const api = spawnAPI({
+            "/tsconfig.json": config,
+            "/src/index.ts": `export const x = 1;`,
+        });
+        try {
+            const snapshot = await api.updateSnapshot({ openProject: "/tsconfig.json" });
+            const project = snapshot.getProject("/tsconfig.json")!;
+            const diags = await project.program.getProgramDiagnostics();
+            assert.deepEqual(diags, [
+                {
+                    fileName: "/tsconfig.json",
+                    ...rangeOf(config, `"bundler"`),
+                    code: 5095,
+                    category: DiagnosticCategory.Error,
+                    text: "Option 'bundler' can only be used when 'module' is set to 'preserve', 'commonjs', or 'es2015' or later.",
+                },
+                {
+                    fileName: "/tsconfig.json",
+                    ...rangeOf(config, `"bundler"`),
+                    code: 5109,
+                    category: DiagnosticCategory.Error,
+                    text: "Option 'moduleResolution' must be set to 'NodeNext' (or left unspecified) when option 'module' is set to 'NodeNext'.",
+                },
+            ]);
+        }
+        finally {
+            await api.close();
+        }
+    });
+
+    test("getGlobalDiagnostics", async () => {
+        const api = spawnAPI({
+            "/tsconfig.json": "{}",
+            "/src/index.ts": `export const x = 1;`,
+        });
+        try {
+            const snapshot = await api.updateSnapshot({ openProject: "/tsconfig.json" });
+            const project = snapshot.getProject("/tsconfig.json")!;
+            const diags = await project.program.getGlobalDiagnostics();
+            assert.deepEqual(diags, []);
+        }
+        finally {
+            await api.close();
+        }
+    });
 });
 
 describe("Checker - getReferencedSymbolsForNode", () => {

--- a/_packages/native-preview/test/async/api.test.ts
+++ b/_packages/native-preview/test/async/api.test.ts
@@ -3518,6 +3518,33 @@ describe("Program - diagnostics", () => {
             await api.close();
         }
     });
+
+    test("getGlobalDiagnostics returns file-less diagnostics from the checker", async () => {
+        const api = spawnAPI({
+            "/tsconfig.json": `{ "compilerOptions": { "noLib": true } }`,
+            "/src/index.ts": `export const x = [1, 2, 3];`,
+        });
+        try {
+            const snapshot = await api.updateSnapshot({ openProject: "/tsconfig.json" });
+            const project = snapshot.getProject("/tsconfig.json")!;
+            const diags = await project.program.getGlobalDiagnostics();
+            // With noLib, the checker reports "Cannot find global type" diagnostics that
+            // are not associated with any source file.
+            assert.ok(diags.length > 0, "expected global diagnostics to be reported");
+            for (const diag of diags) {
+                assert.equal(diag.fileName, undefined);
+                assert.equal(diag.code, 2318);
+                assert.equal(diag.category, DiagnosticCategory.Error);
+            }
+            assert.ok(
+                diags.some(d => d.text === "Cannot find global type 'Array'."),
+                "expected a global diagnostic for the 'Array' type",
+            );
+        }
+        finally {
+            await api.close();
+        }
+    });
 });
 
 describe("Checker - getReferencedSymbolsForNode", () => {

--- a/_packages/native-preview/test/sync/api.test.ts
+++ b/_packages/native-preview/test/sync/api.test.ts
@@ -3446,6 +3446,86 @@ describe("Program - diagnostics", () => {
             api.close();
         }
     });
+
+    test("getBindDiagnostics", () => {
+        const source = `let x = 1;\nlet x = 2;`;
+        const api = spawnAPI({
+            "/tsconfig.json": "{}",
+            "/src/index.ts": source,
+        });
+        try {
+            const snapshot = api.updateSnapshot({ openProject: "/tsconfig.json" });
+            const project = snapshot.getProject("/tsconfig.json")!;
+            const diags = project.program.getBindDiagnostics("/src/index.ts");
+            assert.deepEqual(diags, [
+                {
+                    fileName: "/src/index.ts",
+                    ...rangeOf(source, "x", 0),
+                    code: 2451,
+                    category: DiagnosticCategory.Error,
+                    text: "Cannot redeclare block-scoped variable 'x'.",
+                },
+                {
+                    fileName: "/src/index.ts",
+                    ...rangeOf(source, "x", 1),
+                    code: 2451,
+                    category: DiagnosticCategory.Error,
+                    text: "Cannot redeclare block-scoped variable 'x'.",
+                },
+            ]);
+        }
+        finally {
+            api.close();
+        }
+    });
+
+    test("getProgramDiagnostics", () => {
+        const config = `{ "compilerOptions": { "moduleResolution": "bundler", "module": "nodenext" } }`;
+        const api = spawnAPI({
+            "/tsconfig.json": config,
+            "/src/index.ts": `export const x = 1;`,
+        });
+        try {
+            const snapshot = api.updateSnapshot({ openProject: "/tsconfig.json" });
+            const project = snapshot.getProject("/tsconfig.json")!;
+            const diags = project.program.getProgramDiagnostics();
+            assert.deepEqual(diags, [
+                {
+                    fileName: "/tsconfig.json",
+                    ...rangeOf(config, `"bundler"`),
+                    code: 5095,
+                    category: DiagnosticCategory.Error,
+                    text: "Option 'bundler' can only be used when 'module' is set to 'preserve', 'commonjs', or 'es2015' or later.",
+                },
+                {
+                    fileName: "/tsconfig.json",
+                    ...rangeOf(config, `"bundler"`),
+                    code: 5109,
+                    category: DiagnosticCategory.Error,
+                    text: "Option 'moduleResolution' must be set to 'NodeNext' (or left unspecified) when option 'module' is set to 'NodeNext'.",
+                },
+            ]);
+        }
+        finally {
+            api.close();
+        }
+    });
+
+    test("getGlobalDiagnostics", () => {
+        const api = spawnAPI({
+            "/tsconfig.json": "{}",
+            "/src/index.ts": `export const x = 1;`,
+        });
+        try {
+            const snapshot = api.updateSnapshot({ openProject: "/tsconfig.json" });
+            const project = snapshot.getProject("/tsconfig.json")!;
+            const diags = project.program.getGlobalDiagnostics();
+            assert.deepEqual(diags, []);
+        }
+        finally {
+            api.close();
+        }
+    });
 });
 
 describe("Checker - getReferencedSymbolsForNode", () => {

--- a/_packages/native-preview/test/sync/api.test.ts
+++ b/_packages/native-preview/test/sync/api.test.ts
@@ -3526,6 +3526,33 @@ describe("Program - diagnostics", () => {
             api.close();
         }
     });
+
+    test("getGlobalDiagnostics returns file-less diagnostics from the checker", () => {
+        const api = spawnAPI({
+            "/tsconfig.json": `{ "compilerOptions": { "noLib": true } }`,
+            "/src/index.ts": `export const x = [1, 2, 3];`,
+        });
+        try {
+            const snapshot = api.updateSnapshot({ openProject: "/tsconfig.json" });
+            const project = snapshot.getProject("/tsconfig.json")!;
+            const diags = project.program.getGlobalDiagnostics();
+            // With noLib, the checker reports "Cannot find global type" diagnostics that
+            // are not associated with any source file.
+            assert.ok(diags.length > 0, "expected global diagnostics to be reported");
+            for (const diag of diags) {
+                assert.equal(diag.fileName, undefined);
+                assert.equal(diag.code, 2318);
+                assert.equal(diag.category, DiagnosticCategory.Error);
+            }
+            assert.ok(
+                diags.some(d => d.text === "Cannot find global type 'Array'."),
+                "expected a global diagnostic for the 'Array' type",
+            );
+        }
+        finally {
+            api.close();
+        }
+    });
 });
 
 describe("Checker - getReferencedSymbolsForNode", () => {

--- a/internal/api/proto.go
+++ b/internal/api/proto.go
@@ -140,9 +140,12 @@ const (
 
 	// Diagnostic methods
 	MethodGetSyntacticDiagnostics         Method = "getSyntacticDiagnostics"
+	MethodGetBindDiagnostics              Method = "getBindDiagnostics"
 	MethodGetSemanticDiagnostics          Method = "getSemanticDiagnostics"
 	MethodGetSuggestionDiagnostics        Method = "getSuggestionDiagnostics"
 	MethodGetDeclarationDiagnostics       Method = "getDeclarationDiagnostics"
+	MethodGetProgramDiagnostics           Method = "getProgramDiagnostics"
+	MethodGetGlobalDiagnostics            Method = "getGlobalDiagnostics"
 	MethodGetConfigFileParsingDiagnostics Method = "getConfigFileParsingDiagnostics"
 
 	// Emitter methods
@@ -393,9 +396,12 @@ var unmarshalers = map[Method]func([]byte) (any, error){
 	MethodGetBigIntType:                     unmarshallerFor[GetIntrinsicTypeParams],
 	MethodGetESSymbolType:                   unmarshallerFor[GetIntrinsicTypeParams],
 	MethodGetSyntacticDiagnostics:           unmarshallerFor[GetDiagnosticsParams],
+	MethodGetBindDiagnostics:                unmarshallerFor[GetDiagnosticsParams],
 	MethodGetSemanticDiagnostics:            unmarshallerFor[GetDiagnosticsParams],
 	MethodGetSuggestionDiagnostics:          unmarshallerFor[GetDiagnosticsParams],
 	MethodGetDeclarationDiagnostics:         unmarshallerFor[GetDiagnosticsParams],
+	MethodGetProgramDiagnostics:             unmarshallerFor[GetProjectDiagnosticsParams],
+	MethodGetGlobalDiagnostics:              unmarshallerFor[GetProjectDiagnosticsParams],
 	MethodGetConfigFileParsingDiagnostics:   unmarshallerFor[GetProjectDiagnosticsParams],
 	MethodStartCPUProfile:                   unmarshallerFor[ProfileParams],
 	MethodStopCPUProfile:                    noParams,

--- a/internal/api/session.go
+++ b/internal/api/session.go
@@ -549,12 +549,18 @@ func (s *Session) HandleRequest(ctx context.Context, method string, params json.
 		return s.handleGetIntrinsicType(ctx, parsed.(*GetIntrinsicTypeParams), (*checker.Checker).GetESSymbolType)
 	case string(MethodGetSyntacticDiagnostics):
 		return s.handleGetSyntacticDiagnostics(ctx, parsed.(*GetDiagnosticsParams))
+	case string(MethodGetBindDiagnostics):
+		return s.handleGetBindDiagnostics(ctx, parsed.(*GetDiagnosticsParams))
 	case string(MethodGetSemanticDiagnostics):
 		return s.handleGetSemanticDiagnostics(ctx, parsed.(*GetDiagnosticsParams))
 	case string(MethodGetSuggestionDiagnostics):
 		return s.handleGetSuggestionDiagnostics(ctx, parsed.(*GetDiagnosticsParams))
 	case string(MethodGetDeclarationDiagnostics):
 		return s.handleGetDeclarationDiagnostics(ctx, parsed.(*GetDiagnosticsParams))
+	case string(MethodGetProgramDiagnostics):
+		return s.handleGetProgramDiagnostics(ctx, parsed.(*GetProjectDiagnosticsParams))
+	case string(MethodGetGlobalDiagnostics):
+		return s.handleGetGlobalDiagnostics(ctx, parsed.(*GetProjectDiagnosticsParams))
 	case string(MethodGetConfigFileParsingDiagnostics):
 		return s.handleGetConfigFileParsingDiagnostics(ctx, parsed.(*GetProjectDiagnosticsParams))
 	case string(MethodStartCPUProfile):
@@ -2211,6 +2217,28 @@ func (s *Session) handleGetSyntacticDiagnostics(ctx context.Context, params *Get
 	return NewDiagnosticResponses(diags), nil
 }
 
+// handleGetBindDiagnostics returns bind diagnostics for a file or all files.
+func (s *Session) handleGetBindDiagnostics(ctx context.Context, params *GetDiagnosticsParams) ([]*DiagnosticResponse, error) {
+	ctx = core.WithCheckerLifetime(ctx, core.CheckerLifetimeDiagnostics)
+	sd, err := s.getSnapshotData(params.Snapshot)
+	if err != nil {
+		return nil, err
+	}
+
+	program, err := sd.getProgram(params.Project)
+	if err != nil {
+		return nil, err
+	}
+
+	sourceFile, err := s.resolveOptionalSourceFile(program, params.File)
+	if err != nil {
+		return nil, err
+	}
+
+	diags := program.GetBindDiagnostics(ctx, sourceFile)
+	return NewDiagnosticResponses(diags), nil
+}
+
 // handleGetSemanticDiagnostics returns semantic diagnostics for a file or all files.
 func (s *Session) handleGetSemanticDiagnostics(ctx context.Context, params *GetDiagnosticsParams) ([]*DiagnosticResponse, error) {
 	ctx = core.WithCheckerLifetime(ctx, core.CheckerLifetimeDiagnostics)
@@ -2290,6 +2318,39 @@ func (s *Session) handleGetConfigFileParsingDiagnostics(ctx context.Context, par
 	}
 
 	diags := program.GetConfigFileParsingDiagnostics()
+	return NewDiagnosticResponses(diags), nil
+}
+
+// handleGetProgramDiagnostics returns program-wide diagnostics, including options diagnostics.
+func (s *Session) handleGetProgramDiagnostics(ctx context.Context, params *GetProjectDiagnosticsParams) ([]*DiagnosticResponse, error) {
+	sd, err := s.getSnapshotData(params.Snapshot)
+	if err != nil {
+		return nil, err
+	}
+
+	program, err := sd.getProgram(params.Project)
+	if err != nil {
+		return nil, err
+	}
+
+	diags := program.GetProgramDiagnostics()
+	return NewDiagnosticResponses(diags), nil
+}
+
+// handleGetGlobalDiagnostics returns global (non-file-specific) semantic diagnostics.
+func (s *Session) handleGetGlobalDiagnostics(ctx context.Context, params *GetProjectDiagnosticsParams) ([]*DiagnosticResponse, error) {
+	ctx = core.WithCheckerLifetime(ctx, core.CheckerLifetimeDiagnostics)
+	sd, err := s.getSnapshotData(params.Snapshot)
+	if err != nil {
+		return nil, err
+	}
+
+	program, err := sd.getProgram(params.Project)
+	if err != nil {
+		return nil, err
+	}
+
+	diags := program.GetGlobalDiagnostics(ctx)
 	return NewDiagnosticResponses(diags), nil
 }
 

--- a/internal/api/session.go
+++ b/internal/api/session.go
@@ -49,10 +49,9 @@ type snapshotData struct {
 
 // getProgram looks up a program from a project handle within this snapshot.
 func (sd *snapshotData) getProgram(projectHandle ProjectID) (*compiler.Program, error) {
-	projectName := parseProjectHandle(projectHandle)
-	proj := sd.snapshot.ProjectCollection.GetProjectByPath(projectName)
-	if proj == nil {
-		return nil, fmt.Errorf("%w: project %s not found", ErrClientError, projectName)
+	proj, err := sd.getProject(projectHandle)
+	if err != nil {
+		return nil, err
 	}
 
 	program := proj.GetProgram()
@@ -61,6 +60,16 @@ func (sd *snapshotData) getProgram(projectHandle ProjectID) (*compiler.Program, 
 	}
 
 	return program, nil
+}
+
+// getProject looks up a project from a project handle within this snapshot.
+func (sd *snapshotData) getProject(projectHandle ProjectID) (*project.Project, error) {
+	projectName := parseProjectHandle(projectHandle)
+	proj := sd.snapshot.ProjectCollection.GetProjectByPath(projectName)
+	if proj == nil {
+		return nil, fmt.Errorf("%w: project %s not found", ErrClientError, projectName)
+	}
+	return proj, nil
 }
 
 // nodeHandleFrom creates an index-based node handle (index.kind.path), building a node index table
@@ -2345,12 +2354,26 @@ func (s *Session) handleGetGlobalDiagnostics(ctx context.Context, params *GetPro
 		return nil, err
 	}
 
-	program, err := sd.getProgram(params.Project)
+	proj, err := sd.getProject(params.Project)
 	if err != nil {
 		return nil, err
 	}
 
-	diags := program.GetGlobalDiagnostics(ctx)
+	program := proj.GetProgram()
+	if program == nil {
+		return nil, fmt.Errorf("%w: project has no program", ErrClientError)
+	}
+
+	// Global diagnostics are accumulated lazily by the project's checker pool as
+	// files are checked. Force a full semantic pass so any global (non-file-specific)
+	// diagnostics are produced; otherwise this would return an empty result for
+	// projects using an external checker pool (the typical API case), since
+	// compiler.Program.GetGlobalDiagnostics only reports for the internal pool.
+	program.GetSemanticDiagnostics(ctx, nil)
+
+	diags := core.Filter(proj.GetProjectDiagnostics(ctx), func(d *ast.Diagnostic) bool {
+		return d.File() == nil
+	})
 	return NewDiagnosticResponses(diags), nil
 }
 

